### PR TITLE
Change Vector2/3/4, Quaternion, Plane, Vector<T>, and Vector64/128/256/512<T> to be implemented in managed where trivially possible

### PR DESCRIPTION
**Pull request by tannergooding**
This is a basic prototype for https://github.com/dotnet/runtime/issues/102275 and covers the all vectorized types where the handling can be trivially done in managed (such as by deferring to another intrinsic API).

As part of that, it also removes `MethodImpl(MethodImplOptions.AggressiveInlining)` from various APIs where the method IL size is approx. less than the always inline threshold for RyuJIT (this was estimated by examining a few functions and finding that 2 or less calls with no complex logic typically fits the need).

### CODE REVIEW

**Comment by tannergooding** [comment]
The TL;DR is that I think things look overall positive, but we aren't quite in the state to do this yet.

The biggest issue that I can see is with various intrinsics that expect a constant input. While these are still restricted to import only, the risk of significantly regressing core workloads is significant. If we can enable the support to convert these back to calls in another phase (ideally rationalization like is done for `GT_INTRINSIC`), then this becomes a trivial thing and no longer a concern.

The next concern is just around the number of additional locals and amount of inlining the JIT needs to do. We have tons of zero-ref locals being tracked that are presumably impacting throughput and a non-trivial amount of inlining work being done, leading to higher overall memory usage and lower JIT throughput for methods using such helper APIs.

-------------

Here's a breakdown of some method diffs from `jit-diff.exe diff --diff --pmi --frameworks` for "base" vs this pr.

## `System.SpanHelpers:IndexOfNullCharacter(ulong):int`

We start with seeing that there is a significant number of new single-block inlinees:
```diff
-; 0 inlinees with PGO data; 8 single block inlinees; 4 inlinees without PGO data
+; 0 inlinees with PGO data; 22 single block inlinees; 18 inlinees without PGO data
```

We then see that there is a significant number of additional locals introduced:
```diff
-;* V23 tmp1         [V23    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V24 tmp2         [V24    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V25 tmp3         [V25    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V26 tmp4         [V26    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V27 tmp5         [V27    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V28 tmp6         [V28    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V29 tmp7         [V29    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;* V30 tmp8         [V30    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
-;  V31 tmp9         [V31,T04] ( 10, 11   )     int  ->  rax         "Single return block return value"
-;  V32 cse0         [V32,T03] ( 12, 42   )     ref  ->  rdx         multi-def "CSE #01: aggressive"
-;  V33 rat0         [V33,T09] (  3,  3   )     int  ->  rcx         "ReplaceWithLclVar is creating a new local variable"
+;* V23 tmp1         [V23    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V24 tmp2         [V24    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V25 tmp3         [V25    ] (  0,  0   )  simd32  ->  zero-ref    "spilled call-like call argument"
+;* V26 tmp4         [V26    ] (  0,  0   )  simd32  ->  zero-ref    "spilled call-like call argument"
+;* V27 tmp5         [V27    ] (  0,  0   )  simd64  ->  zero-ref    "spilled call-like call argument"
+;* V28 tmp6         [V28    ] (  0,  0   )  simd64  ->  zero-ref    "spilled call-like call argument"
+;* V29 tmp7         [V29    ] (  0,  0   )  simd32  ->  zero-ref    "spilled call-like call argument"
+;* V30 tmp8         [V30    ] (  0,  0   )  simd32  ->  zero-ref    "spilled call-like call argument"
+;* V31 tmp9         [V31    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V32 tmp10        [V32    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V33 tmp11        [V33    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V34 tmp12        [V34    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V35 tmp13        [V35    ] (  0,  0   )  simd16  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector128`1[ushort]>
+;* V36 tmp14        [V36    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ushort]>
+;* V37 tmp15        [V37    ] (  0,  0   )  simd32  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector256`1[ushort]>
+;* V38 tmp16        [V38    ] (  0,  0   )  simd32  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector256`1[ushort]>
+;* V39 tmp17        [V39    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V40 tmp18        [V40    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V41 tmp19        [V41    ] (  0,  0   )  simd64  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector512`1[ushort]>
+;* V42 tmp20        [V42    ] (  0,  0   )  simd64  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector512`1[ushort]>
+;* V43 tmp21        [V43    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V44 tmp22        [V44    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V45 tmp23        [V45    ] (  0,  0   )  simd32  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector256`1[ushort]>
+;* V46 tmp24        [V46    ] (  0,  0   )  simd32  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector256`1[ushort]>
+;* V47 tmp25        [V47    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V48 tmp26        [V48    ] (  0,  0   )   ubyte  ->  zero-ref    "Inlining Arg"
+;* V49 tmp27        [V49    ] (  0,  0   )  simd16  ->  zero-ref    ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector128`1[ushort]>
+;* V50 tmp28        [V50    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ushort]>
+;  V51 tmp29        [V51,T04] ( 10, 11   )     int  ->  rax         "Single return block return value"
+;  V52 cse0         [V52,T03] ( 12, 42   )     ref  ->  rdx         multi-def "CSE #01: aggressive"
+;  V53 rat0         [V53,T09] (  3,  3   )     int  ->  rcx         "ReplaceWithLclVar is creating a new local variable"
```

The ultimate codegen, however, is identical and most methods fit into this category where there is more inlining and more locals, but ultimately the same codegen.

## `System.Guid:FormatGuidVector128Utf8(System.Guid,ubyte):System.ValueTuple`3[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]`

For this one we again have more inlining:
```diff
-; 0 inlinees with PGO data; 3 single block inlinees; 5 inlinees without PGO data
+; 0 inlinees with PGO data; 22 single block inlinees; 27 inlinees without PGO data
```

And largely a complete change to locals:
```diff
-;  V00 RetBuf       [V00,T00] ( 10,  6   )   byref  ->  rbx         single-def
-;  V01 arg0         [V01,T01] (  3,  6   )   byref  ->  rsi         ld-addr-op single-def
-;  V02 arg1         [V02,T02] (  3,  3   )   ubyte  ->  rdi         single-def
-;  V03 loc0         [V03,T06] (  3,  3   )  simd16  ->  mm6         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V04 loc1         [V04,T04] (  4,  4   )  simd16  ->  mm7         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V05 loc2         [V05,T09] (  4,  2.50)  simd16  ->  mm1         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V06 loc3         [V06    ] (  0,  0   )  simd16  ->  zero-ref    <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V07 loc4         [V07,T12] (  2,  1   )  simd16  ->  mm2         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V08 loc5         [V08,T13] (  2,  1   )  simd16  ->  mm3         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V09 loc6         [V09,T14] (  2,  1   )  simd16  ->  mm1         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V10 loc7         [V10    ] (  0,  0   )  simd16  ->  zero-ref    <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V00 RetBuf       [V00,T04] ( 10,  6   )   byref  ->  rbx         single-def
+;  V01 arg0         [V01,T05] (  3,  6   )   byref  ->  rsi         ld-addr-op single-def
+;  V02 arg1         [V02,T08] (  3,  3   )   ubyte  ->  rdi         single-def
+;  V03 loc0         [V03,T13] (  3,  3   )  simd16  ->  mm6         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V04 loc1         [V04,T12] (  4,  4   )  simd16  ->  mm7         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V05 loc2         [V05    ] (  5,  3.50)  simd16  ->  [rsp+0xB0]  do-not-enreg[XS] addr-exposed <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V06 loc3         [V06    ] (  3,  2   )  simd16  ->  [rsp+0xA0]  do-not-enreg[XS] addr-exposed <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V07 loc4         [V07,T18] (  2,  1   )  simd16  ->  mm6         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V08 loc5         [V08    ] (  2,  1   )  simd16  ->  [rsp+0x90]  do-not-enreg[XS] addr-exposed <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V09 loc6         [V09,T19] (  2,  1   )  simd16  ->  mm0         <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V10 loc7         [V10,T27] (  0,  0   )  simd16  ->  zero-ref    <System.Runtime.Intrinsics.Vector128`1[ubyte]>
 ;* V11 loc8         [V11    ] (  0,  0   )     ref  ->  zero-ref    class-hnd <System.Object>
-;* V12 loc9         [V12    ] (  0,  0   )  simd16  ->  zero-ref    <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V13 loc10        [V13    ] (  0,  0   )  simd16  ->  zero-ref    <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V12 loc9         [V12    ] (  2,  1   )  simd16  ->  [rsp+0x80]  do-not-enreg[XS] addr-exposed <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V13 loc10        [V13    ] (  2,  1   )  simd16  ->  [rsp+0x70]  do-not-enreg[XS] addr-exposed <System.Runtime.Intrinsics.Vector128`1[ubyte]>
 ;* V14 loc11        [V14    ] (  0,  0   )  simd16  ->  zero-ref    ld-addr-op <System.Runtime.Intrinsics.Vector128`1[ubyte]>
 ;  V15 OutArgs      [V15    ] (  1,  1   )  struct (32) [rsp+0x00]  do-not-enreg[XS] addr-exposed "OutgoingArgSpace"
 ;* V16 tmp1         [V16    ] (  0,  0   )     int  ->  zero-ref   
 ;* V17 tmp2         [V17    ] (  0,  0   )  struct (32) zero-ref    "dup spill" <System.ValueTuple`2[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
-;* V18 tmp3         [V18    ] (  0,  0   )  struct (48) zero-ref    ld-addr-op "NewObj constructor temp" <System.ValueTuple`3[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
+;* V18 tmp3         [V18    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
 ;* V19 tmp4         [V19    ] (  0,  0   )  struct (48) zero-ref    ld-addr-op "NewObj constructor temp" <System.ValueTuple`3[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
-;* V20 tmp5         [V20    ] (  0,  0   )     int  ->  zero-ref   
-;  V21 tmp6         [V21,T07] (  3,  3   )  simd16  ->  mm0         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V22 tmp7         [V22,T10] (  2,  2   )  simd16  ->  mm1         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V23 tmp8         [V23,T11] (  2,  2   )  simd16  ->  mm0         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;  V24 tmp9         [V24,T05] (  2,  4   )  simd16  ->  mm1         "impAppendStmt"
-;* V25 tmp10        [V25    ] (  0,  0   )  struct (32) zero-ref    ld-addr-op "NewObj constructor temp" <System.ValueTuple`2[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
-;  V26 tmp11        [V26,T03] (  4,  5   )  simd16  ->  mm0         "spilled call-like call argument"
-;* V27 tmp12        [V27    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V28 tmp13        [V28    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V29 tmp14        [V29    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V30 tmp15        [V30    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V31 tmp16        [V31    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V32 tmp17        [V32,T20] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
-;* V33 tmp18        [V33    ] (  0,  0   )  simd16  ->  zero-ref    "field V17.Item1 (fldOffset=0x0)" P-INDEP
-;* V34 tmp19        [V34    ] (  0,  0   )  simd16  ->  zero-ref    "field V17.Item2 (fldOffset=0x10)" P-INDEP
-;  V35 tmp20        [V35,T15] (  2,  1   )  simd16  ->  mm1         "field V18.Item1 (fldOffset=0x0)" P-INDEP
-;  V36 tmp21        [V36,T16] (  2,  1   )  simd16  ->  mm0         "field V18.Item2 (fldOffset=0x10)" P-INDEP
-;* V37 tmp22        [V37,T21] (  0,  0   )  simd16  ->  zero-ref    "field V18.Item3 (fldOffset=0x20)" P-INDEP
-;  V38 tmp23        [V38,T17] (  2,  1   )  simd16  ->  mm2         "field V19.Item1 (fldOffset=0x0)" P-INDEP
-;  V39 tmp24        [V39,T18] (  2,  1   )  simd16  ->  mm3         "field V19.Item2 (fldOffset=0x10)" P-INDEP
-;  V40 tmp25        [V40,T19] (  2,  1   )  simd16  ->  mm1         "field V19.Item3 (fldOffset=0x20)" P-INDEP
-;* V41 tmp26        [V41    ] (  0,  0   )  simd16  ->  zero-ref    "field V25.Item1 (fldOffset=0x0)" P-INDEP
-;* V42 tmp27        [V42    ] (  0,  0   )  simd16  ->  zero-ref    "field V25.Item2 (fldOffset=0x10)" P-INDEP
-;  V43 cse0         [V43,T08] (  3,  3   )  simd16  ->  mm2         "CSE #01: aggressive"
+;* V20 tmp5         [V20,T25] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V21 tmp6         [V21    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V22 tmp7         [V22    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V23 tmp8         [V23    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V24 tmp9         [V24    ] (  0,  0   )  struct (48) zero-ref    ld-addr-op "NewObj constructor temp" <System.ValueTuple`3[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
+;* V25 tmp10        [V25    ] (  0,  0   )     int  ->  zero-ref   
+;* V26 tmp11        [V26    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;* V27 tmp12        [V27    ] (  0,  0   )  simd16  ->  zero-ref    "spilled call-like call argument"
+;  V28 tmp13        [V28,T14] (  3,  3   )  simd16  ->  mm0         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V29 tmp14        [V29,T16] (  2,  2   )  simd16  ->  mm1         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V30 tmp15        [V30,T17] (  2,  2   )  simd16  ->  mm0         "Inline stloc first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V31 tmp16        [V31,T10] (  3,  6   )  simd16  ->  mm1         "impAppendStmt"
+;* V32 tmp17        [V32    ] (  0,  0   )  struct (32) zero-ref    ld-addr-op "NewObj constructor temp" <System.ValueTuple`2[System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]]>
+;  V33 tmp18        [V33,T11] (  3,  5   )  simd16  ->  mm6         "spilled call-like call argument"
+;* V34 tmp19        [V34    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V35 tmp20        [V35    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V36 tmp21        [V36    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V37 tmp22        [V37    ] (  0,  0   )  simd16  ->  zero-ref    "Inline return value spill temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V38 tmp23        [V38    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V39 tmp24        [V39    ] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[short]>
+;  V40 tmp25        [V40    ] (  2,  4.50)  simd16  ->  [rsp+0x60]  do-not-enreg[XS] addr-exposed ld-addr-op "Inline ldloca(s) first use temp" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V41 tmp26        [V41,T00] (  9, 28.50)     int  ->  rsi         "Inline stloc first use temp"
+;  V42 tmp27        [V42    ] (  2,  9   )  simd16  ->  [rsp+0x50]  do-not-enreg[XS] addr-exposed ld-addr-op "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V43 tmp28        [V43,T01] (  4, 12   )   ubyte  ->  rbp         "Inline stloc first use temp"
+;  V44 tmp29        [V44,T03] (  3, 10   )   ubyte  ->  r14         "Inline stloc first use temp"
+;  V45 tmp30        [V45    ] (  2,  5   )  simd16  ->  [rsp+0x40]  do-not-enreg[XS] addr-exposed ld-addr-op "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V46 tmp31        [V46    ] (  0,  0   )   ubyte  ->  zero-ref    "Inline return value spill temp"
+;  V47 tmp32        [V47,T06] (  3,  8   )     int  ->  rcx        
+;* V48 tmp33        [V48    ] (  0,  0   )   byref  ->  zero-ref    "Inline stloc first use temp"
+;* V49 tmp34        [V49    ] (  0,  0   )   ubyte  ->  zero-ref    "Inline return value spill temp"
+;  V50 tmp35        [V50,T09] (  2,  4   )     int  ->  rcx        
+;* V51 tmp36        [V51    ] (  0,  0   )   byref  ->  zero-ref    "Inline stloc first use temp"
+;  V52 tmp37        [V52,T07] (  3,  8   )     int  ->  rcx        
+;* V53 tmp38        [V53    ] (  0,  0   )   byref  ->  zero-ref    "Inline stloc first use temp"
+;* V54 tmp39        [V54,T26] (  0,  0   )  simd16  ->  zero-ref    "Inlining Arg" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;* V55 tmp40        [V55    ] (  0,  0   )  simd16  ->  zero-ref    "field V17.Item1 (fldOffset=0x0)" P-INDEP
+;* V56 tmp41        [V56    ] (  0,  0   )  simd16  ->  zero-ref    "field V17.Item2 (fldOffset=0x10)" P-INDEP
+;  V57 tmp42        [V57,T20] (  2,  1   )  simd16  ->  mm0         "field V19.Item1 (fldOffset=0x0)" P-INDEP
+;  V58 tmp43        [V58,T21] (  2,  1   )  simd16  ->  mm6         "field V19.Item2 (fldOffset=0x10)" P-INDEP
+;* V59 tmp44        [V59,T28] (  0,  0   )  simd16  ->  zero-ref    "field V19.Item3 (fldOffset=0x20)" P-INDEP
+;  V60 tmp45        [V60,T22] (  2,  1   )  simd16  ->  mm6         "field V24.Item1 (fldOffset=0x0)" P-INDEP
+;  V61 tmp46        [V61,T23] (  2,  1   )  simd16  ->  mm1         "field V24.Item2 (fldOffset=0x10)" P-INDEP
+;  V62 tmp47        [V62,T24] (  2,  1   )  simd16  ->  mm0         "field V24.Item3 (fldOffset=0x20)" P-INDEP
+;* V63 tmp48        [V63    ] (  0,  0   )  simd16  ->  zero-ref    "field V32.Item1 (fldOffset=0x0)" P-INDEP
+;* V64 tmp49        [V64    ] (  0,  0   )  simd16  ->  zero-ref    "field V32.Item2 (fldOffset=0x10)" P-INDEP
+;  V65 tmp50        [V65    ] (  6,  6   )  simd16  ->  [rsp+0x30]  do-not-enreg[XS] addr-exposed "by-value struct argument" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V66 tmp51        [V66    ] (  2,  2   )  simd16  ->  [rsp+0x20]  do-not-enreg[XS] addr-exposed "by-value struct argument" <System.Runtime.Intrinsics.Vector128`1[ubyte]>
+;  V67 cse0         [V67,T15] (  3,  3   )  simd16  ->  mm2         "CSE #01: moderate"
+;  V68 cse1         [V68,T02] (  3, 12   )    long  ->  rdi         "CSE #03: aggressive"
```

However, much of this isn't the same and we get significantly worse codegen:
```diff
-; Lcl frame size = 64
+; Lcl frame size = 224

-        vpshufb  xmm0, xmm6, xmm0
-        vpshufb  xmm1, xmm1, xmmword ptr [reloc @RWD32]
+       vpshufb  xmm6, xmm6, xmm0
+       vmovaps  xmmword ptr [rsp+0xB0], xmm1
+       vmovaps  xmmword ptr [rsp+0xA0], xmm6
+       vpshufb  xmm0, xmm1, xmmword ptr [reloc @RWD32]
+       vmovaps  xmmword ptr [rsp+0xB0], xmm0

-        vpshufb  xmm2, xmm1, xmmword ptr [reloc @RWD48]
+       call     [System.Runtime.Intrinsics.Vector128:Shuffle(System.Runtime.Intrinsics.Vector128`1[ubyte],System.Runtime.Intrinsics.Vector128`1[ubyte]):System.Runtime.Intrinsics.Vector128`1[ubyte]]
```

and so on, with the problem being that since `.As<T, U>` isn't intrinsic, it has to go through inlining. `BitCast` is intrinsic, but forward sub and other considerations happen too late. Since `Shuffle` currently has a pessimization that it requires the value to be constant during import, we fail to do things successfully.

**Comment by tannergooding** [comment]
Reopened now that https://github.com/dotnet/runtime/pull/102702, https://github.com/dotnet/runtime/pull/102827, and https://github.com/dotnet/runtime/pull/102973 has gone in. We should no longer see pessimizations caused from needing constants during import and so we should see fewer diffs overall and ideally be in a position to consider getting this merged and building on it more.

**Comment by tannergooding** [comment]
Things are already looking much better than before.

`minopts` size regressions are expected since it would now require inlining to get the optimized codegen. This is something that needs to be discussed around the overall impact, but I expect its fine since its `T0` code.

`fullopts` has almost no changes and we are doing the right thing for the most part. There are some more complex methods where we run into limits for the default inlining heuristics given that we no longer get the `[Intrinsic]` profitability boost. We could solve this by still treating APIs in `System.Runtime.Intrinsics` as `[Intrinsic]`, regardless of whether or not they're annotated. This gives us the profitability boost without needing to also have the implementation in managed.

There is an assert around `V512.CreateScalar` that needs to be looked at and some mono failures, which at a glance looks to be a legitimate bug in Mono.

File: src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.Extensions.cs [hunk 1]
```diff
@@ -3,27 +3,43 @@
...
 namespace System.Numerics
 {
     public static unsafe partial class Vector
     {
+        /// <summary>Reinterprets a <see cref="Vector4" /> as a new <see cref="Plane" />.</summary>
+        /// <param name="value">The vector to reinterpret.</param>
+        /// <returns><paramref name="value" /> reinterpreted as a new <see cref="Plane" />.</returns>
+        internal static Plane AsPlane(this Vector4 value)
+        {
+#if MONO
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.Extensions.cs:17]: Is this perf problem or functionality problem?

Either way, it is likely going to be fixed as side-effect of https://github.com/dotnet/runtime/issues/102988 .

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/Numerics/Vector4.Extensions.cs:17]: Possibly a functionality problem that I've not dug into yet. The previous tested commit had seen some failures related to Quaternion/Vector4/Plane and their equality tests, but only on Mono.

My current guess is that there's something subtly incorrect in the Mono handling that breaks for SIMD here and it will need a fix before BitCast can be used, but I'd like to try and ensure CoreCLR is clean without regressions before I do any more in depth Mono changes.

**Comment by tannergooding** [comment]
@MihuBot

**Comment by tannergooding** [comment]
One regression is in `System.Drawing.RectangleF:ToVector4()`
```diff
-       vmovups  xmm0, xmmword ptr [rdi]
+       vmovss   xmm0, dword ptr [rdi]
+       vinsertps xmm0, xmm0, dword ptr [rdi+0x04], 16
+       vinsertps xmm0, xmm0, dword ptr [rdi+0x08], 32
+       vinsertps xmm0, xmm0, dword ptr [rdi+0x0C], 48
```

This one is because while we have `fgMorphCombineSIMDFieldStores`, we don't have an `fgMorphCombineSIMDFieldLoads` instead we only had some logic as part of `NI_Vector4_Create` that looked for consecutive field accesses. This one shouldn't be a blocker and can be pretty easily handled in a follow up PR.

---------------

There's a small regression in `System.IO.Hashing.Crc32:UpdateVectorized`:
```diff
 G_M1352_IG02:
        mov      rax, rsi
        mov      ecx, edx
-       vmovups  xmm0, xmmword ptr [rax]
        cmp      ecx, 128
        jge      SHORT G_M1352_IG04
- 						;; size=17 bbWeight=1 PerfScore 5.75
+						;; size=13 bbWeight=1 PerfScore 1.75
 G_M1352_IG03:
-       vmovd    xmm1, rdi
-       vpxor    xmm0, xmm1, xmm0
+       vmovd    xmm0, rdi
+       vpxor    xmm0, xmm0, xmmword ptr [rax]
        jmp      G_M1352_IG07
        align    [0 bytes for IG05]
- 						;; size=13 bbWeight=0.50 PerfScore 2.17
+						;; size=13 bbWeight=0.50 PerfScore 3.50
 G_M1352_IG04:
+       vmovups  xmm0, xmmword ptr [rsi]
        vmovups  xmm1, xmmword ptr [rsi+0x10]
```

Nothing major, just a place where we don't share a load anymore.

-----------------------

In general the JIT ends up needing to create many additional `impAppendStmt`, `Inlining Arg`, `spilled call-like call argument`, and `Inline ldloca(s) first use temp` locals. Most of these end up as `zero-ref` and so there's potential for the JIT to optimize things better around such scenarios to avoid additional or unnecessary overhead.

Most of the corelib regressions are in the xplat APIs that don't have any accelerated implementation today, like `Dot<short>` (lack of handling for multiply since it would have been fairly complex IR to support). Others are cases like `Vector512.OnesComplement` which would be fixed if we had it defer to `op_OnesComplement` rather than `Vector256.OnesComplement` (an easy fix in a follow up PR).

All in all, I think we are in a position where taking this is feasible and it doesn't look to regress any meaningful scenarios, only the already unaccelerated edge cases. -- Mono is being left "as is" for now since their inliner isn't as robust as the RyuJIT one, but we should be able to independently investigate removing the Mono support for the xplat APIs as well so they only need to support the platform specific APIs.

CC. @dotnet/jit-contrib

**Review by EgorBo — APPROVED** [review]
Nice clean up in jit! 👍

**Comment by LoopedBard3** [comment]
Likely regressions:
Windows x64: https://github.com/dotnet/perf-autofiling-issues/issues/36075, https://github.com/dotnet/perf-autofiling-issues/issues/36067
Linux x64: https://github.com/dotnet/perf-autofiling-issues/issues/36054

cc: @tannergooding

**Comment by lewing** [comment]
wasm regressions here https://github.com/dotnet/perf-autofiling-issues/issues/36108

**Comment by tannergooding** [comment]
> Likely regressions:

These should be resolved with https://github.com/dotnet/runtime/pull/103177

> wasm regressions here https://github.com/dotnet/perf-autofiling-issues/issues/36108 (aot)

Little bit surprised that `wasm aot` regressed since the Mono handling hadn't been touched and the AOT support should have already been accelerated for these APIs. Is the support for PackedSIMD separate from the mainline Mono support around MonoLLVM?

> improvements https://github.com/dotnet/perf-autofiling-issues/issues/36093 (interp)

👍, fairly significant improvements at that. I know that `interp` hadn't accelerated the Vector2/3/4 APIs but had accelerated the Vector128 APIs, so lets it do less work and still get the benefits.

**Comment by lewing** [comment]
> > Likely regressions:
> 
> These should be resolved with #103177
> 
> > wasm regressions here [dotnet/perf-autofiling-issues#36108](https://github.com/dotnet/perf-autofiling-issues/issues/36108) (aot)
> 
> Little bit surprised that `wasm aot` regressed since the Mono handling hadn't been touched and the AOT support should have already been accelerated for these APIs. Is the support for PackedSIMD separate from the mainline Mono support around MonoLLVM?
>

I find it surprising too, @radekdoulik can you take a look at our codegen here please

> > improvements [dotnet/perf-autofiling-issues#36093](https://github.com/dotnet/perf-autofiling-issues/issues/36093) (interp)
> 
> 👍, fairly significant improvements at that. I know that `interp` hadn't accelerated the Vector2/3/4 APIs but had accelerated the Vector128 APIs, so lets it do less work and still get the benefits.

Yes, nice improvements

**Comment by matouskozak** [comment]
Possible noticeable size improvements on Mono iOS HelloWorld https://github.com/dotnet/perf-autofiling-issues/issues/35768
